README.md
# axios-cached-dns-resolve

Axios uses node.js dns.lookup to resolve host names.
dns.lookup is synchronous and executes on limited libuv thread pool.
Every axios request will resolve the dns name in kubernetes, openshift, and cloud environments that intentionally set TTL low or to 0 for quick dynamic updates.
The dns resolvers can be overwhelmed with the load.
There is/was a bug in DNS resolutions that manifests as very long dns.lookups in node.js.

From the kubernetes [documentation](https://kubernetes.io/docs/concepts/services-networking/service/#why-not-use-round-robin-dns)

```
Even if apps and libraries did proper re-resolution, the load of every client re-resolving DNS over and over would be difficult to manage.
```

This library uses dns.resolve and can optionally cache resolutions and round-robin among addresses. The cache size is configurable.
If caching is enabled, a background thread will periodically refresh resolutions with dns.resolve rather than every request.
There is an idle TTL that evicts background refresh if an address is no longer being used.
This lib proxies through the OS resolution mechanism which may provide further caching.

## Objectives

  * Async requests - dns resolve vs lookup
  * Fast - local in-app memory cache lookup
  * Fresh - periodically (frequently) updated
  * Constant DNS load/latency vs random load/variable latency
  * Providing statistics and introspection

## Requirements

ECMAScript module (esm), not native esm/.mjs with package.json type: module, requires esm

Node 14+

## Getting started

```console
npm i -S axios-cached-dns-resolve
```

# Usage

```javascript
  import { registerInterceptor } from 'axios-cached-dns-resolve'

  const axiosClient = axios.create(config)

  registerInterceptor(axiosClient)

```
Use axiosClient as normal


## Configuration

```javascript
const config = {
  disabled: process.env.AXIOS_DNS_DISABLE === 'true',
  dnsTtlMs: process.env.AXIOS_DNS_CACHE_TTL_MS || 5000, // when to refresh actively used dns entries (5 sec)
  cacheGraceExpireMultiplier: process.env.AXIOS_DNS_CACHE_EXPIRE_MULTIPLIER || 2, // maximum grace to use entry beyond TTL
  dnsIdleTtlMs: process.env.AXIOS_DNS_CACHE_IDLE_TTL_MS || 1000 * 60 * 60, // when to remove entry entirely if not being used (1 hour)
  backgroundScanMs: process.env.AXIOS_DNS_BACKGROUND_SCAN_MS || 2400, // how frequently to scan for expired TTL and refresh (2.4 sec)
  dnsCacheSize: process.env.AXIOS_DNS_CACHE_SIZE || 100, // maximum number of entries to keep in cache
  // pino logging options
  logging: {
    name: 'axios-cache-dns-resolve',
    // enabled: true,
    level: process.env.AXIOS_DNS_LOG_LEVEL || 'info', // default 'info' others trace, debug, info, warn, error, and fatal
    // timestamp: true,
    prettyPrint: process.env.NODE_ENV === 'DEBUG' || false,
    useLevelLabels: true,
  },
}
```

## Statistics

Statistics are available via 

```javascript
getStats()

{
  "dnsEntries": 4,
  "refreshed": 375679,
  "hits": 128689,
  "misses": 393,
  "idleExpired": 279,
  "errors": 0,
  "lastError": 0,
  "lastErrorTs": 0
}

AND

getDnsCacheEntries()

[
  {
    "host": "foo-service.domain.com",
    "ips": [
      "51.210.235.165",
      "181.73.135.40"
    ],
    "nextIdx": 1,
    "lastUsedTs": 1604151366910,
    "updatedTs": 1604152691039
  },
  ...
]
```

### Express Statistics

```javascript
import { getStats, getDnsCacheEntries } from 'axios-cached-dns-resolve'

router.get('/axios-dns-cache-statistics', getAxiosDnsCacheStatistics)

function getAxiosDnsCacheStatistics(req, resp) {
  resp.json(getStats())
}

router.get('/axios-dns-cache-entries', getAxiosDnsCacheEntries)

function getAxiosDnsCacheEntries(req, resp) {
  resp.json(getDnsCacheEntries())
}
```


Changelog.md
# Changelog

## [3.2.2] - 2022-09-06

### Fixed

- Merged community PR (thanks matrec4) 'Adding a .d.ts file to declare the module for tsnode' [#29](https://github.com/tcollinsworth/axios-cached-dns-resolve/pull/29)


## [3.2.1] - 2022-09-06

### Fixed

- Fixed bug were getDnsCacheEntries was returning generator from lru-cache to instead return array


## [3.2.0] - 2022-09-06

### Changed

- Updated lru-cache to latest version

### Fixed

- Fixed bug were fallback from dns.resolve failure to dns.lookup was not interpreting response array


__tests__/test.js
import ava from 'ava'
import delay from 'delay'
import LRUCache from 'lru-cache'
import axios from 'axios'
import * as axiosCachingDns from '../index.js'

const test = ava.serial

let axiosClient

test.beforeEach(() => {
  axiosCachingDns.config.dnsTtlMs = 1000
  axiosCachingDns.config.dnsIdleTtlMs = 5000
  axiosCachingDns.config.cacheGraceExpireMultiplier = 2
  axiosCachingDns.config.backgroundScanMs = 100

  axiosCachingDns.cacheConfig.ttl = (axiosCachingDns.config.dnsTtlMs * axiosCachingDns.config.cacheGraceExpireMultiplier)

  axiosCachingDns.config.cache = new LRUCache(axiosCachingDns.cacheConfig)

  axiosClient = axios.create({
    timeout: 5000,
    // maxRedirects: 0,
  })

  axiosCachingDns.registerInterceptor(axiosClient)

  axiosCachingDns.startBackgroundRefresh()
  axiosCachingDns.startPeriodicCachePrune()
})

test.after.always(() => {
  axiosCachingDns.config.cache.clear()
})

test('query google with baseURL and relative url', async (t) => {
  axiosCachingDns.registerInterceptor(axios)

  const { data } = await axios.get('/finance', {
    baseURL: 'http://www.google.com',
    // headers: { Authorization: `Basic ${basicauth}` },
  })
  t.truthy(data)
  t.is(1, axiosCachingDns.getStats().dnsEntries)
  t.is(1, axiosCachingDns.getStats().misses)

  const dnsCacheEntries = axiosCachingDns.getDnsCacheEntries()
  /*
   * [
   *   {
   *     host: 'www.google.com',
   *     ips: [ '142.250.190.68' ],
   *     nextIdx: 1,
   *     lastUsedTs: 1665100589485,
   *     updatedTs: 1665100590488
   *   }
   * ]
   */
  t.truthy(Array.isArray(dnsCacheEntries))
  t.truthy(Array.isArray(dnsCacheEntries[0].ips))
  t.truthy(dnsCacheEntries[0].ips[0])
  t.truthy(dnsCacheEntries[0].host)
  t.truthy(dnsCacheEntries[0].lastUsedTs)
  t.truthy(dnsCacheEntries[0].updatedTs)
})

test('query google caches and after idle delay uncached', async (t) => {
  const resp = await axiosClient.get('http://amazon.com')
  t.truthy(resp.data)
  t.truthy(axiosCachingDns.config.cache.get('amazon.com'))
  await delay(6000)
  t.falsy(axiosCachingDns.config.cache.get('amazon.com'))

  const expectedStats = {
    dnsEntries: 0,
    // refreshed: 4, variable
    hits: 0,
    misses: 2,
    idleExpired: 1,
    errors: 0,
    lastError: 0,
    lastErrorTs: 0,
  }

  const stats = axiosCachingDns.getStats()
  delete stats.refreshed
  t.deepEqual(expectedStats, stats)
})

test('query google caches and refreshes', async (t) => {
  await axiosClient.get('http://amazon.com')
  const { updatedTs } = axiosCachingDns.config.cache.get('amazon.com')
  const timeoutTime = Date.now() + 5000
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const dnsEntry = axiosCachingDns.config.cache.get('amazon.com')
    if (!dnsEntry) t.fail('dnsEntry missing or expired')
    // console.log(dnsEntry)
    if (updatedTs !== dnsEntry.updatedTs) break
    if (Date.now() > timeoutTime) t.fail()
    // eslint-disable-next-line no-await-in-loop
    await delay(10)
  }

  const expectedStats = {
    dnsEntries: 1,
    // refreshed: 5, variable
    hits: 0,
    misses: 3,
    idleExpired: 1,
    errors: 0,
    lastError: 0,
    lastErrorTs: 0,
  }

  const stats = axiosCachingDns.getStats()
  delete stats.refreshed
  t.deepEqual(expectedStats, stats)
})

test('query two services, caches and after one idle delay uncached', async (t) => {
  await axiosClient.get('http://amazon.com')

  await axiosClient.get('http://microsoft.com')
  const { lastUsedTs } = axiosCachingDns.config.cache.get('microsoft.com')
  t.is(1, axiosCachingDns.config.cache.get('microsoft.com').nextIdx)

  await axiosClient.get('http://microsoft.com')
  t.is(2, axiosCachingDns.config.cache.get('microsoft.com').nextIdx)

  t.truthy(lastUsedTs < axiosCachingDns.config.cache.get('microsoft.com').lastUsedTs)

  t.is(2, axiosCachingDns.config.cache.size)
  await axiosClient.get('http://microsoft.com')
  t.is(3, axiosCachingDns.config.cache.get('microsoft.com').nextIdx)

  t.falsy(lastUsedTs === axiosCachingDns.config.cache.get('microsoft.com').lastUsedTs)

  t.is(2, axiosCachingDns.config.cache.size)
  await delay(4000)
  t.is(1, axiosCachingDns.config.cache.size)
  await delay(2000)
  t.is(0, axiosCachingDns.config.cache.size)

  const expectedStats = {
    dnsEntries: 0,
    // refreshed: 17, variable
    hits: 2,
    misses: 5,
    idleExpired: 3,
    errors: 0,
    lastError: 0,
    lastErrorTs: 0,
  }

  const stats = axiosCachingDns.getStats()
  delete stats.refreshed
  t.deepEqual(expectedStats, stats)
})

test('validate axios config not altered', async (t) => {
  const baseURL = 'http://microsoft.com'
  const axiosConfig = { baseURL }
  const custAxiosClient = axios.create(axiosConfig)

  axiosCachingDns.registerInterceptor(custAxiosClient)

  await custAxiosClient.get('/')
  t.is(baseURL, axiosConfig.baseURL)
  await custAxiosClient.get('/')
  t.is(baseURL, axiosConfig.baseURL)
})

test('validate axios get config not altered', async (t) => {
  const url = 'http://microsoft.com'
  const custAxiosClient = axios.create()

  const reqConfig = {
    method: 'get',
    url,
  }

  axiosCachingDns.registerInterceptor(custAxiosClient)

  await custAxiosClient.get(url, reqConfig)
  t.is(url, reqConfig.url)
  await custAxiosClient.get(url, reqConfig)
  t.is(url, reqConfig.url)
})

test('validate axios request config not altered', async (t) => {
  const url = 'http://microsoft.com'
  const custAxiosClient = axios.create()

  const reqConfig = {
    method: 'get',
    url,
  }

  axiosCachingDns.registerInterceptor(custAxiosClient)

  await custAxiosClient.request(reqConfig)
  t.is(url, reqConfig.url)
  await custAxiosClient.request(reqConfig)
  t.is(url, reqConfig.url)
})


__tests__/test.mjs
import ava from 'ava'
import delay from 'delay'
import LRUCache from 'lru-cache'
import axios from 'axios'
import * as axiosCachingDns from '../index.mjs'

const test = ava.serial

let axiosClient

test.beforeEach(() => {
  axiosCachingDns.config.dnsTtlMs = 1000
  axiosCachingDns.config.dnsIdleTtlMs = 5000
  axiosCachingDns.config.cacheGraceExpireMultiplier = 2
  axiosCachingDns.config.backgroundScanMs = 100

  axiosCachingDns.cacheConfig.maxAge = (axiosCachingDns.config.dnsTtlMs * axiosCachingDns.config.cacheGraceExpireMultiplier)

  axiosCachingDns.config.cache = new LRUCache(axiosCachingDns.cacheConfig)

  axiosClient = axios.create({
    timeout: 5000,
    // maxRedirects: 0,
  })

  axiosCachingDns.registerInterceptor(axiosClient)

  axiosCachingDns.startBackgroundRefresh()
  axiosCachingDns.startPeriodicCachePrune()
})

test.after.always(() => {
  axiosCachingDns.config.cache.reset()
})

test('query google with baseURL and relative url', async (t) => {
  axiosCachingDns.registerInterceptor(axios)

  const { data } = await axios.get('/finance', {
    baseURL: 'http://www.google.com',
    // headers: { Authorization: `Basic ${basicauth}` },
  })
  t.truthy(data)
  t.is(1, axiosCachingDns.getStats().dnsEntries)
  t.is(1, axiosCachingDns.getStats().misses)
})

test('query google caches and after idle delay uncached', async (t) => {
  const resp = await axiosClient.get('http://amazon.com')
  t.truthy(resp.data)
  t.truthy(axiosCachingDns.config.cache.get('amazon.com'))
  await delay(6000)
  t.falsy(axiosCachingDns.config.cache.get('amazon.com'))

  const expectedStats = {
    dnsEntries: 0,
    // refreshed: 4, variable
    hits: 0,
    misses: 2,
    idleExpired: 1,
    errors: 0,
    lastError: 0,
    lastErrorTs: 0,
  }

  const stats = axiosCachingDns.getStats()
  delete stats.refreshed
  t.deepEqual(expectedStats, stats)
})

test('query google caches and refreshes', async (t) => {
  await axiosClient.get('http://amazon.com')
  const { updatedTs } = axiosCachingDns.config.cache.get('amazon.com')
  const timeoutTime = Date.now() + 5000
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const dnsEntry = axiosCachingDns.config.cache.get('amazon.com')
    if (!dnsEntry) t.fail('dnsEntry missing or expired')
    // console.log(dnsEntry)
    if (updatedTs !== dnsEntry.updatedTs) break
    if (Date.now() > timeoutTime) t.fail()
    // eslint-disable-next-line no-await-in-loop
    await delay(10)
  }

  const expectedStats = {
    dnsEntries: 1,
    // refreshed: 5, variable
    hits: 0,
    misses: 3,
    idleExpired: 1,
    errors: 0,
    lastError: 0,
    lastErrorTs: 0,
  }

  const stats = axiosCachingDns.getStats()
  delete stats.refreshed
  t.deepEqual(expectedStats, stats)
})

test('query two services, caches and after one idle delay uncached', async (t) => {
  await axiosClient.get('http://amazon.com')

  await axiosClient.get('http://microsoft.com')
  const { lastUsedTs } = axiosCachingDns.config.cache.get('microsoft.com')
  t.is(1, axiosCachingDns.config.cache.get('microsoft.com').nextIdx)

  await axiosClient.get('http://microsoft.com')
  t.is(2, axiosCachingDns.config.cache.get('microsoft.com').nextIdx)

  t.truthy(lastUsedTs < axiosCachingDns.config.cache.get('microsoft.com').lastUsedTs)

  t.is(2, axiosCachingDns.config.cache.length)
  await axiosClient.get('http://microsoft.com')
  t.is(3, axiosCachingDns.config.cache.get('microsoft.com').nextIdx)

  t.falsy(lastUsedTs === axiosCachingDns.config.cache.get('microsoft.com').lastUsedTs)

  t.is(2, axiosCachingDns.config.cache.length)
  await delay(4000)
  t.is(1, axiosCachingDns.config.cache.length)
  await delay(2000)
  t.is(0, axiosCachingDns.config.cache.length)

  const expectedStats = {
    dnsEntries: 0,
    // refreshed: 17, variable
    hits: 2,
    misses: 5,
    idleExpired: 3,
    errors: 0,
    lastError: 0,
    lastErrorTs: 0,
  }

  const stats = axiosCachingDns.getStats()
  delete stats.refreshed
  t.deepEqual(expectedStats, stats)
})

test('validate axios config not altered', async (t) => {
  const baseURL = 'http://microsoft.com'
  const axiosConfig = { baseURL }
  const custAxiosClient = axios.create(axiosConfig)

  axiosCachingDns.registerInterceptor(custAxiosClient)

  await custAxiosClient.get('/')
  t.is(baseURL, axiosConfig.baseURL)
  await custAxiosClient.get('/')
  t.is(baseURL, axiosConfig.baseURL)
})

test('validate axios get config not altered', async (t) => {
  const url = 'http://microsoft.com'
  const custAxiosClient = axios.create()

  const reqConfig = {
    method: 'get',
    url,
  }

  axiosCachingDns.registerInterceptor(custAxiosClient)

  await custAxiosClient.get(url, reqConfig)
  t.is(url, reqConfig.url)
  await custAxiosClient.get(url, reqConfig)
  t.is(url, reqConfig.url)
})

test('validate axios request config not altered', async (t) => {
  const url = 'http://microsoft.com'
  const custAxiosClient = axios.create()

  const reqConfig = {
    method: 'get',
    url,
  }

  axiosCachingDns.registerInterceptor(custAxiosClient)

  await custAxiosClient.request(reqConfig)
  t.is(url, reqConfig.url)
  await custAxiosClient.request(reqConfig)
  t.is(url, reqConfig.url)
})


axios-cached-dns-resolve.mjs
/* eslint-disable no-plusplus */
import dns from 'dns'
import URL from 'url'
import net from 'net'
import stringify from 'json-stringify-safe'
import LRUCache from 'lru-cache'
import util from 'util'
import { init as initLogger } from './logging.mjs'

const dnsResolve = util.promisify(dns.resolve)

export const config = {
  disabled: process.env.AXIOS_DNS_DISABLE === 'true',
  dnsTtlMs: process.env.AXIOS_DNS_CACHE_TTL_MS || 5000, // when to refresh actively used dns entries (5 sec)
  cacheGraceExpireMultiplier: process.env.AXIOS_DNS_CACHE_EXPIRE_MULTIPLIER || 2, // maximum grace to use entry beyond TTL
  dnsIdleTtlMs: process.env.AXIOS_DNS_CACHE_IDLE_TTL_MS || 1000 * 60 * 60, // when to remove entry entirely if not being used (1 hour)
  backgroundScanMs: process.env.AXIOS_DNS_BACKGROUND_SCAN_MS || 2400, // how frequently to scan for expired TTL and refresh (2.4 sec)
  dnsCacheSize: process.env.AXIOS_DNS_CACHE_SIZE || 100, // maximum number of entries to keep in cache
  // pino logging options
  logging: {
    name: 'axios-cache-dns-resolve',
    // enabled: true,
    level: process.env.AXIOS_DNS_LOG_LEVEL || 'info', // default 'info' others trace, debug, info, warn, error, and fatal
    // timestamp: true,
    prettyPrint: process.env.NODE_ENV === 'DEBUG' || false,
    useLevelLabels: true,
  },
  cache: undefined,
}

export const cacheConfig = {
  max: config.dnsCacheSize,
  maxAge: (config.dnsTtlMs * config.cacheGraceExpireMultiplier), // grace for refresh
}

export const stats = {
  dnsEntries: 0,
  refreshed: 0,
  hits: 0,
  misses: 0,
  idleExpired: 0,
  errors: 0,
  lastError: 0,
  lastErrorTs: 0,
}

let log
let backgroundRefreshId
let cachePruneId

init()

export function init() {
  log = initLogger(config.logging)

  if (config.cache) return

  config.cache = new LRUCache(cacheConfig)

  startBackgroundRefresh()
  startPeriodicCachePrune()
  cachePruneId = setInterval(() => config.cache.prune(), config.dnsIdleTtlMs)
}

export function startBackgroundRefresh() {
  if (backgroundRefreshId) clearInterval(backgroundRefreshId)
  backgroundRefreshId = setInterval(backgroundRefresh, config.backgroundScanMs)
}

export function startPeriodicCachePrune() {
  if (cachePruneId) clearInterval(cachePruneId)
  cachePruneId = setInterval(() => config.cache.prune(), config.dnsIdleTtlMs)
}

export function getStats() {
  stats.dnsEntries = config.cache.length
  return stats
}

export function getDnsCacheEntries() {
  return config.cache.values()
}

// const dnsEntry = {
//   host: 'www.amazon.com',
//   ips: [
//     '52.54.40.141',
//     '34.205.98.207',
//     '3.82.118.51',
//   ],
//   nextIdx: 0,
//   lastUsedTs: 1555771516581, Date.now()
//   updatedTs: 1555771516581,
// }

export function registerInterceptor(axios) {
  if (config.disabled || !axios || !axios.interceptors) return // supertest
  axios.interceptors.request.use(async (reqConfig) => {
    try {
      let url
      if (reqConfig.baseURL) {
        url = URL.parse(reqConfig.baseURL)
      } else {
        url = URL.parse(reqConfig.url)
      }

      if (net.isIP(url.hostname)) return reqConfig // skip

      reqConfig.headers.Host = url.hostname // set hostname in header

      url.hostname = await getAddress(url.hostname)
      delete url.host // clear hostname

      if (reqConfig.baseURL) {
        reqConfig.baseURL = URL.format(url)
      } else {
        reqConfig.url = URL.format(url)
      }
    } catch (err) {
      recordError(err, `Error getAddress, ${err.message}`)
    }

    return reqConfig
  })
}

export async function getAddress(host) {
  let dnsEntry = config.cache.get(host)
  if (dnsEntry) {
    ++stats.hits
    dnsEntry.lastUsedTs = Date.now()
    // eslint-disable-next-line no-plusplus
    const ip = dnsEntry.ips[dnsEntry.nextIdx++ % dnsEntry.ips.length] // round-robin
    config.cache.set(host, dnsEntry)
    return ip
  }
  ++stats.misses
  if (log.isLevelEnabled('debug')) log.debug(`cache miss ${host}`)

  const ips = await dnsResolve(host)
  dnsEntry = {
    host,
    ips,
    nextIdx: 0,
    lastUsedTs: Date.now(),
    updatedTs: Date.now(),
  }
  // eslint-disable-next-line no-plusplus
  const ip = dnsEntry.ips[dnsEntry.nextIdx++ % dnsEntry.ips.length] // round-robin
  config.cache.set(host, dnsEntry)
  return ip
}

let backgroundRefreshing = false
export async function backgroundRefresh() {
  if (backgroundRefreshing) return // don't start again if currently iterating slowly
  backgroundRefreshing = true
  try {
    config.cache.forEach(async (value, key) => {
      try {
        if (value.updatedTs + config.dnsTtlMs > Date.now()) {
          return // continue/skip
        }
        if (value.lastUsedTs + config.dnsIdleTtlMs <= Date.now()) {
          ++stats.idleExpired
          config.cache.del(key)
          return // continue
        }

        const ips = await dnsResolve(value.host)
        value.ips = ips
        value.updatedTs = Date.now()
        config.cache.set(key, value)
        ++stats.refreshed
      } catch (err) {
        // best effort
        recordError(err, `Error backgroundRefresh host: ${key}, ${stringify(value)}, ${err.message}`)
      }
    })
  } catch (err) {
    // best effort
    recordError(err, `Error backgroundRefresh, ${err.message}`)
  } finally {
    backgroundRefreshing = false
  }
}

function recordError(err, errMesg) {
  ++stats.errors
  stats.lastError = err
  stats.lastErrorTs = new Date().toISOString()
  log.error(err, errMesg)
}
/* eslint-enable no-plusplus */


axios-cached-dns-resolve.js
/* eslint-disable no-plusplus */
import dns from 'dns'
import URL from 'url'
import net from 'net'
import stringify from 'json-stringify-safe'
import LRUCache from 'lru-cache'
import util from 'util'
import { init as initLogger } from './logging.js'

const dnsResolve = util.promisify(dns.resolve)
const dnsLookup = util.promisify(dns.lookup)

export const config = {
  disabled: process.env.AXIOS_DNS_DISABLE === 'true',
  dnsTtlMs: process.env.AXIOS_DNS_CACHE_TTL_MS || 5000, // when to refresh actively used dns entries (5 sec)
  cacheGraceExpireMultiplier: process.env.AXIOS_DNS_CACHE_EXPIRE_MULTIPLIER || 2, // maximum grace to use entry beyond TTL
  dnsIdleTtlMs: process.env.AXIOS_DNS_CACHE_IDLE_TTL_MS || 1000 * 60 * 60, // when to remove entry entirely if not being used (1 hour)
  backgroundScanMs: process.env.AXIOS_DNS_BACKGROUND_SCAN_MS || 2400, // how frequently to scan for expired TTL and refresh (2.4 sec)
  dnsCacheSize: process.env.AXIOS_DNS_CACHE_SIZE || 100, // maximum number of entries to keep in cache
  // pino logging options
  logging: {
    name: 'axios-cache-dns-resolve',
    // enabled: true,
    level: process.env.AXIOS_DNS_LOG_LEVEL || 'info', // default 'info' others trace, debug, info, warn, error, and fatal
    // timestamp: true,
    prettyPrint: process.env.NODE_ENV === 'DEBUG' || false,
    formatters: {
      level(label/* , number */) {
        return { level: label }
      },
    },
  },
  cache: undefined,
}

export const cacheConfig = {
  max: config.dnsCacheSize,
  ttl: (config.dnsTtlMs * config.cacheGraceExpireMultiplier), // grace for refresh
}

export const stats = {
  dnsEntries: 0,
  refreshed: 0,
  hits: 0,
  misses: 0,
  idleExpired: 0,
  errors: 0,
  lastError: 0,
  lastErrorTs: 0,
}

let log
let backgroundRefreshId
let cachePruneId

init()

export function init() {
  log = initLogger(config.logging)

  if (config.cache) return

  config.cache = new LRUCache(cacheConfig)

  startBackgroundRefresh()
  startPeriodicCachePrune()
  cachePruneId = setInterval(() => config.cache.purgeStale(), config.dnsIdleTtlMs)
}

export function startBackgroundRefresh() {
  if (backgroundRefreshId) clearInterval(backgroundRefreshId)
  backgroundRefreshId = setInterval(backgroundRefresh, config.backgroundScanMs)
}

export function startPeriodicCachePrune() {
  if (cachePruneId) clearInterval(cachePruneId)
  cachePruneId = setInterval(() => config.cache.purgeStale(), config.dnsIdleTtlMs)
}

export function getStats() {
  stats.dnsEntries = config.cache.size
  return stats
}

export function getDnsCacheEntries() {
  return Array.from(config.cache.values())
}

// const dnsEntry = {
//   host: 'www.amazon.com',
//   ips: [
//     '52.54.40.141',
//     '34.205.98.207',
//     '3.82.118.51',
//   ],
//   nextIdx: 0,
//   lastUsedTs: 1555771516581, Date.now()
//   updatedTs: 1555771516581,
// }

export function registerInterceptor(axios) {
  if (config.disabled || !axios || !axios.interceptors) return // supertest
  axios.interceptors.request.use(async (reqConfig) => {
    try {
      let url
      if (reqConfig.baseURL) {
        url = URL.parse(reqConfig.baseURL)
      } else {
        url = URL.parse(reqConfig.url)
      }

      if (net.isIP(url.hostname)) return reqConfig // skip

      reqConfig.headers.Host = url.hostname // set hostname in header

      url.hostname = await getAddress(url.hostname)
      delete url.host // clear hostname

      if (reqConfig.baseURL) {
        reqConfig.baseURL = URL.format(url)
      } else {
        reqConfig.url = URL.format(url)
      }
    } catch (err) {
      recordError(err, `Error getAddress, ${err.message}`)
    }

    return reqConfig
  })
}

export async function getAddress(host) {
  let dnsEntry = config.cache.get(host)
  if (dnsEntry) {
    ++stats.hits
    dnsEntry.lastUsedTs = Date.now()
    // eslint-disable-next-line no-plusplus
    const ip = dnsEntry.ips[dnsEntry.nextIdx++ % dnsEntry.ips.length] // round-robin
    config.cache.set(host, dnsEntry)
    return ip
  }
  ++stats.misses
  if (log.isLevelEnabled('debug')) log.debug(`cache miss ${host}`)

  const ips = await resolve(host)
  dnsEntry = {
    host,
    ips,
    nextIdx: 0,
    lastUsedTs: Date.now(),
    updatedTs: Date.now(),
  }
  // eslint-disable-next-line no-plusplus
  const ip = dnsEntry.ips[dnsEntry.nextIdx++ % dnsEntry.ips.length] // round-robin
  config.cache.set(host, dnsEntry)
  return ip
}

let backgroundRefreshing = false
export async function backgroundRefresh() {
  if (backgroundRefreshing) return // don't start again if currently iterating slowly
  backgroundRefreshing = true
  try {
    config.cache.forEach(async (value, key) => {
      try {
        if (value.updatedTs + config.dnsTtlMs > Date.now()) {
          return // continue/skip
        }
        if (value.lastUsedTs + config.dnsIdleTtlMs <= Date.now()) {
          ++stats.idleExpired
          config.cache.delete(key)
          return // continue
        }

        const ips = await resolve(value.host)
        value.ips = ips
        value.updatedTs = Date.now()
        config.cache.set(key, value)
        ++stats.refreshed
      } catch (err) {
        // best effort
        recordError(err, `Error backgroundRefresh host: ${key}, ${stringify(value)}, ${err.message}`)
      }
    })
  } catch (err) {
    // best effort
    recordError(err, `Error backgroundRefresh, ${err.message}`)
  } finally {
    backgroundRefreshing = false
  }
}

/**
 *
 * @param host
 * @returns {*[]}
 */
async function resolve(host) {
  let ips
  try {
    ips = await dnsResolve(host)
    throw new Error('test')
  } catch (e) {
    let lookupResp = await dnsLookup(host, { all: true }) // pass options all: true for all addresses
    lookupResp = extractAddresses(lookupResp)
    if (!Array.isArray(lookupResp) || lookupResp.length < 1) throw new Error(`fallback to dnsLookup returned no address ${host}`)
    ips = lookupResp
  }
  return ips
}

// dns.lookup
// ***************** { address: '142.250.190.68', family: 4 }
// , { all: true } /***************** [ { address: '142.250.190.68', family: 4 } ]

function extractAddresses(lookupResp) {
  if (!Array.isArray(lookupResp)) throw new Error('lookup response did not contain array of addresses')
  return lookupResp.filter((e) => e.address != null).map((e) => e.address)
}

function recordError(err, errMesg) {
  ++stats.errors
  stats.lastError = err
  stats.lastErrorTs = new Date().toISOString()
  log.error(err, errMesg)
}
/* eslint-enable no-plusplus */
